File: GuideObjC/GuideObjC/Objects/ReversedEdgeDownloader.m
//
//  ReversedEdgeDownloader.m
//  GuideObjC
//
//  Created by LuzanovRoman on 22.12.16.
//  Copyright © 2016 eigengraph. All rights reserved.
//

#import "ReversedEdgeDownloader.h"

@implementation ReversedEdgeDownloader

- (NSObject *)last {
    return self.graphObjects.firstObject;
}

- (void)addObjects:(NSArray *)objects totalCount:(NSInteger)count {
    if (objects) {
        NSMutableArray *indexPaths = [NSMutableArray array];
        
        for (NSInteger i = 0; i < objects.count; i++) {
            [indexPaths addObject:[NSIndexPath indexPathForRow:i inSection:1]];
        }
        self.totalCount = count;
        
        for (NSObject * object in objects) {
            [self.graphObjects insertObject:object atIndex:0];
        }
        [self.tableView beginUpdates];
        [self.tableView insertRowsAtIndexPaths:indexPaths withRowAnimation:UITableViewRowAnimationNone];
        [self.tableView reloadRowsAtIndexPaths:@[[NSIndexPath indexPathForRow:0 inSection:0]] withRowAnimation:UITableViewRowAnimationNone];
        [self.tableView endUpdates];
    }
}

- (void)replaceObject:(NSObject *)object {
    if (object) {
        NSInteger index = [self indexOfObject:object];
        
        if (index == NSNotFound) {
            return;
        }
        [self.graphObjects replaceObjectAtIndex:index withObject:object];
        
        if (self.tableView) {
            if ([self.delegate respondsToSelector:@selector(willUpdateGraphObject:)]) {
                [self.delegate willUpdateGraphObject:object];
            }
            [self.tableView beginUpdates];
            [self.tableView reloadRowsAtIndexPaths:@[[NSIndexPath indexPathForRow:index inSection:1]] withRowAnimation:UITableViewRowAnimationNone];
            [self.tableView endUpdates];
            
            if ([self.delegate respondsToSelector:@selector(didUpdateGraphObject:)]) {
                [self.delegate didUpdateGraphObject:object];
            }
        }
    }
}

- (void)insertObject:(NSObject *)object atIndex:(NSInteger)index {
    if (object) {
        // Check if object is already in the list
        if ([self indexOfObject:object] != NSNotFound) {
            return;
        }
        NSInteger insertIndex = self.graphObjects.count - index;
        self.totalCount += 1;
        [self.graphObjects insertObject:object atIndex:insertIndex];
        [self.tableView beginUpdates];
        [self.tableView insertRowsAtIndexPaths:@[[NSIndexPath indexPathForRow:insertIndex inSection:1]] withRowAnimation:UITableViewRowAnimationNone];
        [self.tableView endUpdates];
    }
}

- (void)deleteObject:(NSObject *)object {
    NSInteger index = [self.graphObjects indexOfObject:object];
    
    if (index == NSNotFound) {
        return;
    }
    self.totalCount -= 1;
    [self.graphObjects removeObjectAtIndex:index];
    
    [self.tableView beginUpdates];
    [self.tableView deleteRowsAtIndexPaths:@[[NSIndexPath indexPathForRow:index inSection:1]] withRowAnimation:UITableViewRowAnimationNone];
    [self.tableView endUpdates];
}
@end

File: GuideSwift/GuideSwift/Objects/ReversedEdgeDownloader.swift
//
//  ReversedEdgeDownloader.swift
//  GuideSwift
//
//  Created by LuzanovRoman on 22.12.16.
//  Copyright © 2016 eigengraph. All rights reserved.
//

import UIKit

class ReversedEdgeDownloader<T: NSObject>: EdgeDownloader<T> {
    
    override var last: T? {
        get {
            return graphObjects.first
        }
    }
    
    override func add(objects: [T]?, totalCount count: Int) {
        guard let theObjects = objects else { return }
        
        let addAnimated = graphObjects.count > 0
        var indexPaths = [IndexPath]()
        
        for i in 0..<theObjects.count {
            indexPaths.append(IndexPath(row: i, section: 1))
        }
        totalCount = count
        
        for object in theObjects {
            graphObjects.insert(object, at: 0)
        }
        
        guard let tv = tableView else { return }
        
        if addAnimated {
            tv.beginUpdates()
            tv.insertRows(at: indexPaths, with: .none)
            tv.reloadRows(at: [IndexPath(row: 0, section: 0)], with: .none)
            tv.endUpdates()
        }
        else {
            tv.reloadData()
        }
    }
    
    override func replace(object: T) {
        guard let index = indexOf(object: object) else { return }
        graphObjects.replaceSubrange(index..<index + 1, with: [object])
        
        guard let tv = tableView else { return }
        delegate?.willUpdate?(graphObject: object)
        tv.beginUpdates()
        tv.reloadRows(at: [IndexPath(row: index, section: 1)], with: .none)
        tv.endUpdates()
        delegate?.didUpdate?(graphObject: object)
    }
    
    override func insert(object: T?, at index: Int) {
        guard let theObject = object, let objectId = theObject.value(forKey: "id") as? String else { return }
        
        // Check if object is already in the list
        if let _ = indexOfObject(withId: objectId) {
            return
        }
        let insertIndex = graphObjects.count - index
        totalCount += 1
        graphObjects.insert(theObject, at: insertIndex)
        
        guard let tv = tableView else { return }
        tv.beginUpdates()
        tv.insertRows(at: [IndexPath(row: insertIndex, section: 1)], with: .none)
        tv.endUpdates()
    }
    
    override func delete(object: T?) {
        guard let theObject = object, let index = graphObjects.index(of: theObject) else { return }
        totalCount -= 1
        graphObjects.remove(at: index)
        
        guard let tv = tableView else { return }
        tv.beginUpdates()
        tv.deleteRows(at: [IndexPath(row: index, section: 1)], with: .none)
        tv.endUpdates()
    }
}
